File: bumi/src/http/gateway.ts
/**
 * Gateway API compiler — HttpRouteMatch → Matcher<HttpRequest, A>.
 *
 * Translates Gateway API-style route configuration into xuma Matcher trees.
 * Pure TypeScript types mirroring the Gateway API spec (no k8s dependency).
 */

import { type Matcher, MatcherError, matcherFromPredicate } from "../matcher.ts";
import { SinglePredicate, andPredicate, orPredicate } from "../predicate.ts";
import type { Predicate } from "../predicate.ts";
import { ExactMatcher, PrefixMatcher, RegexMatcher } from "../string-matchers.ts";
import { HeaderInput, MethodInput, PathInput, QueryParamInput } from "./inputs.ts";
import type { HttpRequest } from "./request.ts";

export interface HttpPathMatch {
	readonly type: "Exact" | "PathPrefix" | "RegularExpression";
	readonly value: string;
}

export interface HttpHeaderMatch {
	readonly type: "Exact" | "RegularExpression";
	readonly name: string;
	readonly value: string;
}

export interface HttpQueryParamMatch {
	readonly type: "Exact" | "RegularExpression";
	readonly name: string;
	readonly value: string;
}

/** Gateway API HttpRouteMatch — all conditions ANDed together. */
export interface HttpRouteMatch {
	readonly path?: HttpPathMatch;
	readonly method?: string;
	readonly headers?: readonly HttpHeaderMatch[];
	readonly queryParams?: readonly HttpQueryParamMatch[];
}

/** Compile a single route match into a Matcher. */
export function compileRouteMatch<A>(
	routeMatch: HttpRouteMatch,
	action: A,
): Matcher<HttpRequest, A> {
	return matcherFromPredicate(routeMatchToPredicate(routeMatch), action);
}

/**
 * Compile multiple route matches (ORed) into a single Matcher.
 *
 * Substituting a catch-all for an empty array was never spec behaviour. xDS is
 * explicit: "if no matcher above matched and this field is not populated, the
 * match will be considered unsuccessful" — an empty list is a *no-match*. The
 * config path already honours that; only this convenience layer disagreed with
 * the engine underneath it.
 *
 * Not fixed by copying the loader, because there `onNoMatch` is config the
 * operator wrote, while here it is an argument: `([], "allow", "deny")` and
 * `([], "deny", "allow")` are opposite outcomes from the same empty input. An
 * empty list is also almost never written on purpose — it is a config that
 * failed to load, or a filter that removed every rule.
 *
 * @throws {MatcherError} if `matches` is empty, or any entry has no conditions.
 */
export function compileRouteMatches<A>(
	matches: readonly HttpRouteMatch[],
	action: A,
	onNoMatch?: A,
): Matcher<HttpRequest, A> {
	if (matches.length === 0) {
		throw new MatcherError(
			"no route matches, which would match every request. Use compileCatchAll() " +
				"if that is intended, or onNoMatch for a default route.",
		);
	}
	const predicates = matches.map((m) => routeMatchToPredicate(m));
	return matcherFromPredicate(orPredicate(predicates, catchAll()), action, onNoMatch);
}

/**
 * Build a matcher that matches every request.
 *
 * The explicit form of what `compileRouteMatches` now refuses to do by
 * accident. A catch-all is a legitimate route; it just has to be asked for, and
 * it is greppable when someone later asks why a gate admits everything.
 */
export function compileCatchAll<A>(action: A): Matcher<HttpRequest, A> {
	return matcherFromPredicate(catchAll(), action);
}

/** A catch-all predicate that matches any HTTP request. */
function catchAll(): Predicate<HttpRequest> {
	return new SinglePredicate(new PathInput(), new PrefixMatcher(""));
}

function routeMatchToPredicate(rm: HttpRouteMatch): Predicate<HttpRequest> {
	const predicates: SinglePredicate<HttpRequest>[] = [];

	if (rm.path !== undefined) {
		predicates.push(compilePathMatch(rm.path));
	}
	if (rm.method !== undefined) {
		predicates.push(new SinglePredicate(new MethodInput(), new ExactMatcher(rm.method)));
	}
	for (const h of rm.headers ?? []) {
		predicates.push(compileHeaderMatch(h));
	}
	for (const q of rm.queryParams ?? []) {
		predicates.push(compileQueryParamMatch(q));
	}

	// An empty conjunction is vacuously true, so andPredicate would hand back
	// catchAll(). Reaching this is rarely deliberate: HttpRouteMatch is an
	// interface with every field optional and no runtime schema at all, so a
	// JSON config saying `pathPrefix` where it meant `path` arrives here with
	// nothing set and nothing to signal it.
	//
	// This is also the only moment the mistake is visible. After substitution
	// the predicate is PrefixMatcher("") on the path and is indistinguishable
	// from a deliberate catch-all, which is why Matcher.validate() cannot catch
	// it and never could.
	if (predicates.length === 0) {
		throw new MatcherError(
			"HttpRouteMatch has no conditions, so it matches every request — check " +
				"for a misspelled field. Use compileCatchAll() if a catch-all is intended.",
		);
	}

	return andPredicate(predicates, catchAll());
}

function compilePathMatch(pm: HttpPathMatch): SinglePredicate<HttpRequest> {
	switch (pm.type) {
		case "Exact":
			return new SinglePredicate(new PathInput(), new ExactMatcher(pm.value));
		case "PathPrefix":
			return new SinglePredicate(new PathInput(), new PrefixMatcher(pm.value));
		case "RegularExpression":
			return new SinglePredicate(new PathInput(), new RegexMatcher(pm.value));
		default:
			throw new MatcherError(`Unknown path match type: ${pm.type}`);
	}
}

function compileHeaderMatch(hm: HttpHeaderMatch): SinglePredicate<HttpRequest> {
	const input = new HeaderInput(hm.name);
	switch (hm.type) {
		case "Exact":
			return new SinglePredicate(input, new ExactMatcher(hm.value));
		case "RegularExpression":
			return new SinglePredicate(input, new RegexMatcher(hm.value));
		default:
			throw new MatcherError(`Unknown header match type: ${hm.type}`);
	}
}

function compileQueryParamMatch(qm: HttpQueryParamMatch): SinglePredicate<HttpRequest> {
	const input = new QueryParamInput(qm.name);
	switch (qm.type) {
		case "Exact":
			return new SinglePredicate(input, new ExactMatcher(qm.value));
		case "RegularExpression":
			return new SinglePredicate(input, new RegexMatcher(qm.value));
		default:
			throw new MatcherError(`Unknown query param match type: ${qm.type}`);
	}
}

File: puma/src/xuma/http/_gateway.py
"""Gateway API compiler — HttpRouteMatch -> Matcher[HttpRequest, A].

Translates Gateway API-style route configuration into xuma Matcher trees.
Pure Python types mirroring the Gateway API spec (no k8s dependency).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Literal

from xuma._matcher import Matcher, MatcherError, matcher_from_predicate
from xuma._predicate import Predicate, SinglePredicate, and_predicate, or_predicate
from xuma._string_matchers import ExactMatcher, PrefixMatcher, RegexMatcher
from xuma.http._inputs import HeaderInput, MethodInput, PathInput, QueryParamInput

if TYPE_CHECKING:
    from xuma.http._request import HttpRequest


def _catch_all() -> Predicate[HttpRequest]:
    """A catch-all predicate that matches any HTTP request."""
    return SinglePredicate(PathInput(), PrefixMatcher(""))


@dataclass(frozen=True, slots=True)
class HttpPathMatch:
    """Gateway API path match specification."""

    type: Literal["Exact", "PathPrefix", "RegularExpression"]
    value: str


@dataclass(frozen=True, slots=True)
class HttpHeaderMatch:
    """Gateway API header match specification."""

    type: Literal["Exact", "RegularExpression"]
    name: str
    value: str


@dataclass(frozen=True, slots=True)
class HttpQueryParamMatch:
    """Gateway API query parameter match specification."""

    type: Literal["Exact", "RegularExpression"]
    name: str
    value: str


@dataclass(frozen=True, slots=True)
class HttpRouteMatch:
    """Gateway API HttpRouteMatch — configuration for route matching.

    All conditions within a single HttpRouteMatch are ANDed together.
    Multiple HttpRouteMatch entries are ORed (via compile_route_matches).
    """

    path: HttpPathMatch | None = None
    method: str | None = None
    headers: list[HttpHeaderMatch] = field(default_factory=list)
    query_params: list[HttpQueryParamMatch] = field(default_factory=list)

    def compile[A](self, action: A) -> Matcher[HttpRequest, A]:
        """Compile this route match into a Matcher with the given action."""
        return matcher_from_predicate(self.to_predicate(), action)

    def to_predicate(self) -> Predicate[HttpRequest]:
        """Convert this route match to a predicate tree."""
        predicates: list[Predicate[HttpRequest]] = []

        if self.path is not None:
            predicates.append(_compile_path_match(self.path))

        if self.method is not None:
            predicates.append(SinglePredicate(MethodInput(), ExactMatcher(self.method)))

        for header_match in self.headers:
            predicates.append(_compile_header_match(header_match))

        for query_match in self.query_params:
            predicates.append(_compile_query_param_match(query_match))

        # An empty conjunction is vacuously true, so and_predicate would hand
        # back _catch_all(). Reaching this is rarely deliberate: every field is
        # optional and nothing rejects an unknown one, so a config saying
        # `pathPrefix:` where it meant `path:` arrives here with nothing set and
        # no signal that anything went wrong.
        #
        # This is also the only moment the mistake is visible. After
        # substitution the predicate is PrefixMatcher("") on the path, which is
        # indistinguishable from a deliberate catch-all — which is why
        # Matcher.validate() cannot catch it and never could.
        if not predicates:
            msg = (
                "HttpRouteMatch has no conditions, so it matches every request "
                "— check for a misspelled field. Use compile_catch_all() if a "
                "catch-all is intended."
            )
            raise MatcherError(msg)

        return and_predicate(predicates, _catch_all())


def compile_route_matches[A](
    matches: list[HttpRouteMatch],
    action: A,
    on_no_match: A | None = None,
) -> Matcher[HttpRequest, A]:
    """Compile multiple HttpRouteMatch entries into a single Matcher.

    Multiple matches are ORed together per Gateway API semantics.

    Raises:
        MatcherError: If the list is empty, or if any entry has no conditions.
            Both are catch-alls.
    """
    # Substituting a catch-all here was never spec behaviour. xDS is explicit:
    # "if no matcher above matched and this field is not populated, the match
    # will be considered unsuccessful" — an empty list is a *no-match*. The
    # config path already gets this right; only this convenience layer
    # disagreed with the engine underneath it.
    #
    # Not fixed by copying the loader, because there on_no_match is config the
    # operator wrote, while here it is an argument: ([], "allow", "deny") and
    # ([], "deny", "allow") are opposite outcomes from the same empty input.
    # An empty list is also almost never written on purpose — it is a config
    # that failed to load, or a filter that removed every rule.
    if not matches:
        msg = (
            "no route matches, which would match every request. Use "
            "compile_catch_all() if that is intended, or on_no_match for a "
            "default route."
        )
        raise MatcherError(msg)

    predicates = [m.to_predicate() for m in matches]
    return matcher_from_predicate(
        or_predicate(predicates, _catch_all()),
        action,
        on_no_match,
    )


def compile_catch_all[A](action: A) -> Matcher[HttpRequest, A]:
    """Build a matcher that matches every request.

    The explicit form of what compile_route_matches() now refuses to do by
    accident. A catch-all is a legitimate route; it just has to be asked for,
    and it is greppable when someone later asks why a gate admits everything.
    """
    return matcher_from_predicate(_catch_all(), action)


def _compile_path_match(path_match: HttpPathMatch) -> SinglePredicate[HttpRequest]:
    """Compile a path match to a predicate."""
    match path_match.type:
        case "Exact":
            return SinglePredicate(PathInput(), ExactMatcher(path_match.value))
        case "PathPrefix":
            return SinglePredicate(PathInput(), PrefixMatcher(path_match.value))
        case "RegularExpression":
            return SinglePredicate(PathInput(), RegexMatcher(path_match.value))
        case _:
            msg = f"Unknown path match type: {path_match.type}"
            raise ValueError(msg)


def _compile_header_match(header_match: HttpHeaderMatch) -> SinglePredicate[HttpRequest]:
    """Compile a header match to a predicate."""
    input_ = HeaderInput(header_match.name)
    match header_match.type:
        case "Exact":
            return SinglePredicate(input_, ExactMatcher(header_match.value))
        case "RegularExpression":
            return SinglePredicate(input_, RegexMatcher(header_match.value))
        case _:
            msg = f"Unknown header match type: {header_match.type}"
            raise ValueError(msg)


def _compile_query_param_match(query_match: HttpQueryParamMatch) -> SinglePredicate[HttpRequest]:
    """Compile a query param match to a predicate."""
    input_ = QueryParamInput(query_match.name)
    match query_match.type:
        case "Exact":
            return SinglePredicate(input_, ExactMatcher(query_match.value))
        case "RegularExpression":
            return SinglePredicate(input_, RegexMatcher(query_match.value))
        case _:
            msg = f"Unknown query param match type: {query_match.type}"
            raise ValueError(msg)
